File: backend/models/HealthMetric.js
const mongoose = require('mongoose');

const healthMetricSchema = new mongoose.Schema({
  patientId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Patient',
    required: true
  },
  
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  
  deviceId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Device'
  },
  
  metricType: {
    type: String,
    required: true,
    enum: [
      'steps', 'heart_rate', 'blood_pressure', 'blood_glucose',
      'weight', 'height', 'bmi', 'body_temperature',
      'oxygen_saturation', 'sleep_duration', 'calories_burned',
      'water_intake', 'respiratory_rate'
    ]
  },
  
  value: {
    type: mongoose.Schema.Types.Mixed,
    required: true
  },
  
  unit: {
    type: String,
    required: true
  },
  
  timestamp: {
    type: Date,
    required: true,
    default: Date.now
  },
  
  source: {
    type: String,
    enum: ['device', 'manual', 'calculated', 'imported'],
    default: 'manual'
  },
  
  deviceName: String,
  
  qualityScore: {
    type: Number,
    min: 0,
    max: 100,
    default: 100
  },
  
  isAbnormal: {
    type: Boolean,
    default: false
  },
  
  accuracy: {
    type: Number,
    min: 0,
    max: 100
  },
  
  notes: String,
  tags: [String],
  
  isVerified: {
    type: Boolean,
    default: false
  },
  
  verifiedBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Doctor'
  },
  
  verificationDate: Date
}, {
  timestamps: true
});

healthMetricSchema.index({ patientId: 1, metricType: 1, timestamp: -1 });
healthMetricSchema.index({ userId: 1 });
healthMetricSchema.index({ deviceId: 1 });
healthMetricSchema.index({ timestamp: 1 });
healthMetricSchema.index({ metricType: 1, isAbnormal: 1 });

healthMetricSchema.index({ patientId: 1, timestamp: -1, metricType: 1 });

healthMetricSchema.pre('save', function(next) {
  if (!this.unit) {
    const unitMap = {
      steps: 'steps',
      heart_rate: 'bpm',
      blood_pressure: 'mmHg',
      blood_glucose: 'mg/dL',
      weight: 'kg',
      height: 'cm',
      body_temperature: '°C',
      oxygen_saturation: '%',
      sleep_duration: 'hours',
      calories_burned: 'kcal',
      water_intake: 'ml',
      respiratory_rate: 'breaths/min'
    };
    
    this.unit = unitMap[this.metricType] || 'unknown';
  }
  
  if (this.value && !this.isAbnormal) {
    this.detectAbnormal();
  }
  
  next();
});

healthMetricSchema.methods.detectAbnormal = function() {
  const normalRanges = {
    heart_rate: { min: 60, max: 100 },
    blood_pressure_systolic: { min: 90, max: 120 },
    blood_pressure_diastolic: { min: 60, max: 80 },
    blood_glucose: { min: 70, max: 140 },
    body_temperature: { min: 36.1, max: 37.2 },
    oxygen_saturation: { min: 95, max: 100 },
    respiratory_rate: { min: 12, max: 20 }
  };
  
  if (this.metricType === 'blood_pressure' && typeof this.value === 'object') {
    const systolic = this.value.systolic;
    const diastolic = this.value.diastolic;
    
    this.isAbnormal = systolic < 90 || systolic > 140 || 
                      diastolic < 60 || diastolic > 90;
  } else if (normalRanges[this.metricType]) {
    const range = normalRanges[this.metricType];
    this.isAbnormal = this.value < range.min || this.value > range.max;
  }
};

healthMetricSchema.statics.getLatestMetrics = function(patientId, metricTypes) {
  const query = { patientId };
  
  if (metricTypes && metricTypes.length > 0) {
    query.metricType = { $in: metricTypes };
  }
  
  return this.find(query)
    .sort({ timestamp: -1 })
    .limit(50)
    .exec();
};

healthMetricSchema.statics.getMetricTrend = function(patientId, metricType, startDate, endDate) {
  return this.find({
    patientId,
    metricType,
    timestamp: { $gte: startDate, $lte: endDate }
  })
  .sort({ timestamp: 1 })
  .select('value timestamp')
  .exec();
};

const HealthMetric = mongoose.model('HealthMetric', healthMetricSchema);

module.exports = HealthMetric;

File: frontend/src/auth2_flutter/lib/features/data/domain/entities/health_metric.dart
class HealthMetric {
  final String id;
  final String patientId;
  final String userId;
  final String? deviceId;
  final String metricType;
  final dynamic value;
  final String unit;
  final DateTime timestamp;
  final String source;
  final String? deviceName;
  final int qualityScore;
  final bool isAbnormal;
  final double? accuracy;
  final String? notes;
  final List<String> tags;

  HealthMetric({
    required this.id,
    required this.patientId,
    required this.userId,
    this.deviceId,
    required this.metricType,
    required this.value,
    required this.unit,
    required this.timestamp,
    required this.source,
    this.deviceName,
    required this.qualityScore,
    required this.isAbnormal,
    this.accuracy,
    this.notes,
    required this.tags,
  });

  factory HealthMetric.fromJson(Map<String, dynamic> json) {
    return HealthMetric(
      id: json['_id']?.toString() ?? '',
      patientId: json['patientId'] is Map
          ? json['patientId']['_id'].toString()
          : json['patientId'].toString(),
      userId: json['userId'] is Map
          ? json['userId']['_id'].toString()
          : json['userId'].toString(),
      deviceId: json['deviceId'] == null
          ? null
          : (json['deviceId'] is Map
              ? json['deviceId']['_id'].toString()
              : json['deviceId'].toString()),
      metricType: json['metricType'] ?? '',
      value: json['value'],
      unit: json['unit'] ?? '',
      timestamp: json['timestamp'] != null
          ? DateTime.parse(json['timestamp'])
          : DateTime.now(),
      source: json['source'] ?? 'manual',
      deviceName: json['deviceName'],
      qualityScore: json['qualityScore'] != null
          ? (json['qualityScore'] is int
              ? json['qualityScore']
              : int.tryParse(json['qualityScore'].toString()) ?? 100)
          : 100,
      isAbnormal: json['isAbnormal'] ?? false,
      accuracy: json['accuracy'] != null
          ? (json['accuracy'] is double
              ? json['accuracy']
              : double.tryParse(json['accuracy'].toString()))
          : null,
      notes: json['notes'],
      tags: List<String>.from(json['tags'] ?? []),
    );
  }
}
